sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -767,29 +745,201 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
+            numChunks = numChunks == 0 ? 1 : numChunks;
+
+            BlobDownloadAsyncResponse initialResponse = setupTuple3.getT3();
+            return Flux.range(0, numChunks)
+                .flatMap(chunkNum -> {
+                    // The first chunk was retrieved during setup.
+                    if (chunkNum == 0) {
+                        return writeBodyToFile(initialResponse, file, 0, finalParallelTransferOptions, progressLock,
+                            totalProgress);
+                    }
+
+                    // Calculate whether we need a full chunk or something smaller because we are at the end.
+                    long chunkSizeActual = Math.min(finalParallelTransferOptions.getBlockSize(),
+                        newCount - (chunkNum * finalParallelTransferOptions.getBlockSize()));
+                    BlobRange chunkRange = new BlobRange(
+                        rangeReal.getOffset() + (chunkNum * finalParallelTransferOptions.getBlockSize()),
+                        chunkSizeActual);
+
+                    // Make the download call.
+                    return this.downloadWithResponse(chunkRange, downloadRetryOptions, realConditions,
+                        rangeGetContentMd5, null)
+                        .subscribeOn(Schedulers.elastic())
+                        .flatMap(response ->
+                            writeBodyToFile(response, file, chunkNum, finalParallelTransferOptions, progressLock,
+                                totalProgress));
+                })
+                // All the headers will be the same, so we just pick the last one.
+                .last();
+        });
     }
 
-    private List<BlobRange> sliceBlobRange(BlobRange blobRange, Integer blockSize) {
-        if (blockSize == null) {
-            blockSize = BLOB_DEFAULT_DOWNLOAD_BLOCK_SIZE;
+    private int calculateNumBlocks(long dataSize, long blockLength) {
+        // Can successfully cast to an int because MaxBlockSize is an int, which this expression must be less than.
+        int numBlocks = toIntExact(dataSize / blockLength);
+        // Include an extra block for trailing data.
+        if (dataSize % blockLength != 0) {
+            numBlocks++;
         }
-        long offset = blobRange.getOffset();
-        long length = blobRange.getCount();
-        List<BlobRange> chunks = new ArrayList<>();
-        for (long pos = offset; pos < offset + length; pos += blockSize) {
-            long count = blockSize;
-            if (pos + count > offset + length) {
-                count = offset + length - pos;
+        return numBlocks;
+    }
+
+    /*
+    Download the first chunk. Construct a Mono which will emit the total count for calculating the number of chunks,
+    access conditions containing the etag to lock on, and the response from downloading the first chunk.
+     */
+    private Mono<Tuple3<Long, BlobRequestConditions, BlobDownloadAsyncResponse>> getSetupMono(BlobRange r,
```

Comment:
Could we use full names for parameters, `range` instead of `r` and `options` instead of `o`? It'll make reading the code later a little less cryptic.